### ISSUE TITLE
Why does Conformer add x twice for each block (ffn1, ffn2, conv, and mhsa)?

### ISSUE THREAD

**Issue by lunixbochs** [flashlight/flashlight#582]
Edit from the future: it's also doing this for the mhsa and conv blocks, which I missed while researching the original issue

During the forward pass:
https://github.com/flashlight/flashlight/blob/master/flashlight/fl/contrib/modules/Conformer.cpp#L175-L181
https://github.com/flashlight/flashlight/blob/master/flashlight/fl/contrib/modules/Conformer.cpp#L187-L193

```
ffn1 = x + ffn1(x)
x = x + 0.5 * ffn1
ffn2 = x + ffn2(x)
x = x + 0.5 * ffn2
```

This basically weighs x at 1.5 each ffn step instead of 1.0

The other conformer implementations I've found don't do this (they only scale down ffn, they don't double-add x)
- https://github.com/sooftware/conformer/blob/main/conformer/encoder.py#L79
- https://github.com/lucidrains/conformer/blob/0e91f0323089cc00fbde3e112a984d94d0af09b8/conformer/conformer.py#L190

**Comment by lunixbochs** [comment]
I'm also confused about the pair of "pointwise conv" steps just being linear layers in flashlight:
https://github.com/flashlight/flashlight/blob/master/flashlight/fl/contrib/modules/Conformer.cpp#L40-L42

**Comment by tlikhomanenko** [comment]
Hey!

> This basically weighs x at 1.5 each ffn step instead of 1.0

Thanks for pointing! I am not sure what implementation is the right one. For our implementation I directly follows the paper https://arxiv.org/pdf/2005.08100.pdf. Probably we can remove this weight from 1.5 to 1. - if you have comparison between the two versions, let me know.

> I'm also confused about the pair of "pointwise conv" steps just being linear layers in flashlight:

Pointwise convolutions is conv with kernel 1x1 which equivalent to apply linear layer on channel dim. So we can either call conv 1x1 or apply linear layer (we do the latest).

**Comment by lunixbochs** [comment]
> For our implementation I directly follows the paper

I don't see the 1.5x weight in the paper, it still seems to me like adding x twice may be a bug. Here's their Conformer Block diagram, which only adds x once at each step for a standard residual connection:

<img width="415" alt="Screen Shot 2021-05-05 at 7 05 54 PM" src="https://user-images.githubusercontent.com/117642/117231853-eca76f00-add4-11eb-87c2-873668f760ae.png">

I don't have the resources to prioritize an experiment on this, intuitively it seems to me like it would hurt the network by over-amplifying the residual values. If my math is right, the residual path will end up as something like `431439x` larger across 16 conformer blocks (via `x * (1.5 ** 32)`) for no reason.

> Pointwise convolutions is conv with kernel 1x1 which equivalent to apply linear layer on channel dim

I didn't see it at first but I ran this experiment and now I see the equivalence of pointwise conv and linear in the channel dimension. Thanks!

```
import torch
import torch.nn as nn

with torch.no_grad():
    conv = nn.Conv1d(1, 2, kernel_size=1, bias=False)
    conv.weight = nn.Parameter(torch.tensor([1, 2], dtype=torch.float64).reshape(2, 1, 1))
    fc = nn.Linear(1, 2, bias=False)
    fc.weight = nn.Parameter(torch.tensor([[1], [2]], dtype=torch.float64))
    x = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.float64)
    x = x.reshape(3, 1, 3)
    print(x)
    print('conv')
    print(conv(x))
    x = x.reshape(3, 3, 1)
    print('fc')
    print(fc(x).transpose(1, 2))
```

**Comment by tlikhomanenko** [comment]
> I don't see the 1.5x weight in the paper, it still seems to me like adding x twice may be a bug. Here's their Conformer Block diagram, which only adds x once at each step for a standard residual connection:

Look at the figure 4 where ffn layer definition is given, so we are doing exactly it in the block https://github.com/flashlight/flashlight/blob/master/flashlight/fl/contrib/modules/Conformer.cpp#L187-L193.

**Comment by lunixbochs** [comment]
They use that skip connection notation on every block diagram, see also figures 2 and 3.

The extra add ultimately makes no sense to me due the point of residual skip connections, which is to pass the input through unmodified (not to multiply the original value by 400,000). If you applied it for figures 2 and 3 as well you'd be multiplying the original x by 186 billion, which I think could cause precision issues even at fp32. (I think the 400k multiply could cause precision issues at fp16 which only has +-64k of precision).

I will also note that my 28M conformer that outperforms the RASR models was trained at full precision, and I haven't been able to reproduce it at half precision. This could be related?

**Comment by lunixbochs** [comment]
Here's the bug as code to demonstrate the impact:
```
import math

def fn(x):
    return math.sin(x) + 1

def model1(x):
    print(f'model1({x})')
    for i in range(16):
        ffn1 = fn(x)
        x    = x + ffn1
        ffn2 = fn(x)
        x    = x + ffn2
        print(f'i={i+1:<2} x={x:.6f} ffn1={ffn1:.6f} ffn2={ffn2:.6f}')
    return x
        
def model2(x):
    print(f'model2({x})')
    for i in range(16):
        ffn1 = x + fn(x)
        x    = x + 0.5 * ffn1
        ffn2 = x + fn(x)
        x    = x + 0.5 * ffn2
        print(f'i={i+1:<2} x={x:>14.6f} ffn1={ffn1:>14.6f} ffn2={ffn2:>14.6f}')
    return x
        
model1(1)
model2(1)
```

Output:

```
model1(1)
i=1  x=4.137107 ffn1=1.841471 ffn2=1.295636
i=2  x=4.382678 ffn1=0.160961 ffn2=0.084610
i=3  x=4.474347 ffn1=0.053864 ffn2=0.037805
i=4  x=4.524482 ffn1=0.028198 ffn2=0.021936
i=5  x=4.556551 ffn1=0.017603 ffn2=0.014467
i=6  x=4.578979 ffn1=0.012118 ffn2=0.010310
i=7  x=4.595608 ffn1=0.008886 ffn2=0.007743
i=8  x=4.608460 ffn1=0.006811 ffn2=0.006041
i=9  x=4.618706 ffn1=0.005396 ffn2=0.004850
i=10 x=4.627076 ffn1=0.004385 ffn2=0.003984
i=11 x=4.634046 ffn1=0.003637 ffn2=0.003334
i=12 x=4.639945 ffn1=0.003067 ffn2=0.002832
i=13 x=4.645005 ffn1=0.002623 ffn2=0.002436
i=14 x=4.649393 ffn1=0.002269 ffn2=0.002119
i=15 x=4.653238 ffn1=0.001984 ffn2=0.001861
i=16 x=4.656634 ffn1=0.001749 ffn2=0.001647

model2(1)
i=1  x=      4.461118 ffn1=      2.841471 ffn2=      4.080764
i=2  x=     10.766859 ffn1=      4.492521 ffn2=      8.118963
i=3  x=     24.525078 ffn1=     10.792901 ffn2=     16.723536
i=4  x=     55.682264 ffn1=     24.954127 ffn2=     37.360244
i=5  x=    126.426009 ffn1=     55.920259 ffn2=     85.567231
i=6  x=    286.683423 ffn1=    128.116598 ffn2=    192.398230
i=7  x=    645.865051 ffn1=    286.967119 ffn2=    431.396139
i=8  x=   1454.189986 ffn1=    645.900686 ffn2=    970.749182
i=9  x=   3273.942748 ffn1=   1455.549188 ffn2=   2183.956337
i=10 x=   7367.433547 ffn1=   3275.335114 ffn2=   4911.646483
i=11 x=  16577.375035 ffn1=   7368.045258 ffn2=  11051.837717
i=12 x=  37300.417417 ffn1=  16579.098907 ffn2=  24866.985859
i=13 x=  83926.624290 ffn1=  37301.133534 ffn2=  55951.280212
i=14 x= 188837.205417 ffn1=  83928.478199 ffn2= 125892.684055
i=15 x= 424885.020047 ffn1= 188838.913936 ffn2= 283256.715323
i=16 x= 955991.968523 ffn1= 424885.703950 ffn2= 637328.193003
```

**Comment by tlikhomanenko** [comment]
Yep, makes sense for me now. I know why I did this extra res connection - their 1/2 in formula (1). Now I think you are right, as for conv I interpreted formula (1) as the res connection from the plot (so no extra res connection here) but not for ffn as they have 1/2 in formula (1) but not on the figure plot. About the fp32 vs fp16 - could be, but I have a doubt, because there is layer norm in the block at the end, so this should normalize all things to the proper scales I guess. Let me discuss with the team how to elaborate this bug so that we can reuse our released conformer model and will send the fix then (or feel free to send PR as you found the bug :) )

**Comment by lunixbochs** [comment]
I completely missed earlier that the same extra skip connection is happening in `mhsa()` and `conv()`, which makes more sense given your earlier reasoning.

https://github.com/flashlight/flashlight/blob/9990ad421f016b4a72b0b6a9942081823df989d3/flashlight/fl/contrib/modules/Conformer.cpp#L159
https://github.com/flashlight/flashlight/blob/9990ad421f016b4a72b0b6a9942081823df989d3/flashlight/fl/contrib/modules/Conformer.cpp#L140

Here the issue is adding `input`, which acts as a `2x` factor for conv/mhsa (vs the `1.5x` factor for the ffn block).

I've had a colleague test the impact of the extra `0.5*x` connection while training with AMP. Early estimates are that it did hurt the model's accuracy by maybe 5-10% to have the extra `x` added for ffn. I imagine the impact of conv/mhsa could be even more due to the larger `2x` factor.

---

I've thought about how to fix this. You could add a constructor argument to the Conformer block with a default value e.g. `version=1`, and do a version check to see if these additions should happen. Newly created Conformer blocks via `SequentialBuilder` can have version 2, but the existing Conformer models would deserialize with the default `version=1` and retain their old behavior.

**Comment by joazoa** [comment]
These are some test results to measure the impact of Lunixbochs' patch: 
With patch vs baseline
loss:    0.22884	0.27922
train-TER:  0.82	1.01
train-WER:  3.51	4.23
DS1-loss:    1.10995	1.21176
DS1-TER:  3.38		3.71
DS1-WER: 12.48		13.60
DS2-loss:    0.43201	0.44926
DS2-TER:  0.91		1.01
DS2-WER:  3.52		3.88 
DS3-loss:    0.51346	0.57324 
DS3-TER:  1.47		1.64
DS3-WER:  4.99		5.57
DS4-loss:    0.62241	0.69094
DS4-TER:  1.82		2.01 
DS4-WER:  5.65		6.26 
DS5-loss 2.50190	2.44709 
DS5-TER:  7.24		7.22
DS5-WER: 20.14		21.02

28m conformer, saug disabled, mixed precision=true. Identical nupdates and dynamicbatchsize.

Patch:
--- a/flashlight/fl/contrib/modules/Conformer.cpp
+++ b/flashlight/fl/contrib/modules/Conformer.cpp
@@ -173,7 +173,6 @@ std::vector<Variable> Conformer::forward(const std::vector<Variable>& input) {
   auto x = input[0];
   // apply first feed-forward module
   auto ffn1 =
-      x +
       dropout(
           (*w12_)(dropout(
               fl::swish((*w11_)(((*norm1_)(x)).as(x.type())), 1.), pDropout)),
@@ -185,7 +184,6 @@ std::vector<Variable> Conformer::forward(const std::vector<Variable>& input) {
   x = x + f * conv(x);
   // apply second feed-forward module
   auto ffn2 =
-      x +
       dropout(
           (*w22_)(dropout(
               fl::swish((*w21_)(((*norm2_)(x)).as(x.type())), 1.), pDropout)),

**Comment by tlikhomanenko** [comment]
All nice catches and results! Let me test this patch for rasr model, as will have models result will merge the patch. Thanks a lot for all work you did @lunixbochs @joazoa !

**Comment by joazoa** [comment]
These are some results from all 3 patches: (baseline versus new)

loss:    0.25212 | 0.21177
train-TER:  0.90 | 0.77 
train-WER:  3.84 | 3.18 
DS1:    1.28120 | 1.10865 
DS1-TER:  3.74 | 3.14 
DS1-WER: 13.48 | 11.49 
DS2-loss:    0.46202 | 0.47701 
DS2-TER:  0.99 | 0.86 
DS2-WER:  3.78 | 3.21 
DS3-loss:    0.57905 | 0.47689 
DS3-TER:  1.58 | 1.36 
DS3-WER:  5.36 | 4.51 
DS4-loss:    0.70196 | 0.56444 
DS4-TER:  1.98 | 1.68 
DS4-WER:  6.13 | 5.16 
DS5-loss:    2.66550 | 2.30005 
DS5-TER:  7.07 | 6.67 
DS5-WER: 20.68 | 18.44

**Comment by tlikhomanenko** [comment]
merged
